openssl-dynamic/src/main/c/ocsp.c

```diff
@@ -0,0 +1,358 @@
+
+    return response;
+}
+
+/**
+ * The client uses -1 for errors which is distinctly different from the server's error code.
+ *
+ * @see netty_ocsp_callback
+ */
+static int netty_ocsp_client_callback(SSL *ssl, void *arg) {
+    return netty_ocsp_callback(-1, ssl, arg);
+}
+
+/**
+ * The server uses SSL_TLSEXT_ERR_ALERT_FATAL for errors (it's a non-negative number).
+ *
+ * @see netty_ocsp_callback
+ */
+static int netty_ocsp_server_callback(SSL *ssl, void *arg) {
+    return netty_ocsp_callback(SSL_TLSEXT_ERR_ALERT_FATAL, ssl, arg);
+}
+
+
+/**
+ * Returns true is OCSP stapling is available.
+ */
+TCN_IMPLEMENT_CALL(jboolean, SSLContext, isOcspSupported)(TCN_STDARGS) {
+    return JNI_TRUE;
+}
+
+/**
+ * Installs the OCSP stapling callback.
+ */
+TCN_IMPLEMENT_CALL(jobject, SSLContext, newOcspCallback)(TCN_STDARGS, jlong ctx, jboolean client, jobject callback) {
+    long error = -1L;
+
+    tcn_ssl_ctxt_t *c = J2P(ctx, tcn_ssl_ctxt_t *);
+    if (c == NULL) {
+        tcn_ThrowException(e, "ctx is null");
+        return NULL;
+    }
+
+    //
+    // ATTENTION: We need to create a global reference of the callback.
+    //
+    jweak callbackRef = (*e)->NewWeakGlobalRef(e, callback);
+
+    error = SSL_CTX_set_tlsext_status_arg(c->ctx, callbackRef);
+    if (error <= 0L) {
+        netty_ocsp_free_callback(e, NULL, callbackRef);
+
+        tcn_ThrowException(e, "SSL_CTX_set_tlsext_status_arg() failed");
+        return NULL;
+    }
+
+    // The client and server callbacks use slightly different return
+    // values. We need therefore two different functions. We could use
+    // a struct as an argument for SSL_CTX_set_tlsext_status_arg() as
+    // well but that involve a lot more work.
+
+    if (client) {
+        error = SSL_CTX_set_tlsext_status_cb(c->ctx, netty_ocsp_client_callback);
+
+    } else {
+        error = SSL_CTX_set_tlsext_status_cb(c->ctx, netty_ocsp_server_callback);
+    }
+
+    if (error <= 0L) {
+        netty_ocsp_free_callback(e, NULL, callbackRef);
+
+        tcn_ThrowException(e, "SSL_CTX_set_tlsext_status_cb() failed");
+        return NULL;
+    }
+
+    return callbackRef;
+}
+
+/**
+ * Releases the OCSP callback.
+ */
+TCN_IMPLEMENT_CALL(void, SSLContext, freeOcspCallback)(TCN_STDARGS, jlong ctx, jobject callback) {
+    tcn_ssl_ctxt_t *c = J2P(ctx, tcn_ssl_ctxt_t *);
+    netty_ocsp_free_callback(e, c, callback);
+}
+
+/**
+ * Enables or disables OCSP stapling for the given SSLEngine.
+ */
+TCN_IMPLEMENT_CALL(jboolean, SSL, setStatusType)(TCN_STDARGS, jlong ssl, jint type) {
+    SSL *ssl_ = J2P(ssl, SSL *);
+    if (ssl_ == NULL) {
+        tcn_ThrowException(e, "ssl is null");
+        return JNI_FALSE;
+    }
+
+    long error = SSL_set_tlsext_status_type(ssl_, type);
+    return (error == 1L) ? JNI_TRUE : JNI_FALSE;
+}
+
+/**
+ * Server: Sets the stapled OCSP record bytes for the given SSLEngine.
+ */
+TCN_IMPLEMENT_CALL(jboolean, SSL, setOcspStaple)(TCN_STDARGS, jlong ssl, jbyteArray response) {
```

Comment:
> Is this in regards to the setter or also getter?

Yip. The getter and setter.

> tcnative's existing code that deals direct memory seems to revolve around BIOs and that doesn't seem like a good fit

BIOs should only be necessary if there are expected by OpenSSL. However in this case we are just trying to have memory which can be accessed by java and JNI, and direct byte buffer is a good fit now. See [writeToSSL](https://github.com/netty/netty-tcnative/blob/1.1.33/openssl-dynamic/src/main/c/ssl.c#L1448) in netty-tcnative and its usage [in netty](https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/ssl/ReferenceCountedOpenSslEngine.java#L382).

> In Java land you'll most likely end up working with something like Bouncycastle's OCSPReq and OCSPResp classes to work with the staples and that's byte[] based.

We can always also support a `byte[]` if necessary, and/or we can also do the copying from direct memory to `byte[]` in java.